### DISCUSSION TITLE
Customizing layout root element based on current route

### DISCUSSION THREAD

**Discussion by ptasheq** [solidjs/solid-router#379] [Q&A]
Hello,

I am wondering if there is an elegant way to customize the layout component passed as `root` prop to `Router` depending on which `Route` is active.

It feels very natural to me that some component (its markup) e.g. header repeats over all pages, but for one attribute, so I may place it in the layout component and access router state/definition to set customizable prop. Still, I don't see a solution for that on current solid-router version.

Thank you in advance

**Comment by hawkerboy7** [comment]
Not sure if I understand you correctly. If you'd add a code example of what you have and would like it might make it easier.

Maybe you can pick up ideas from my current setup (simplified and untested):
```javascript

export default function() {

	const Main = (props) => {
		return (<>
			<Header/>
			{props.children}
			<Footer/>
		</>)
	}

	const Routes = () => <Router root={Main}><Pages/></Router>

	render(Routes, document.getElementById("main"))
}
```

The `Pages` component then contains the `Routes` so each navigation the `Header` and `Footer` component remains the same.
```javascript
export default function Pages() {
	return (<>
		<Route path="/" component={Home} />
		<Route path="/test" component={Test} />
	</>)
}
```

The DOM then would look something like this:
```html
<body>
	<div id="main">
		<header></header>
		<div class="home"></div>
		<footer></footer>
	</div>
</body>
```

Edit: Upon reading you question multiple times I think you mean to adjust e.g. the `Header` in the example above based on which path/component is currently active?
So when you have the component `Home` active the header is for example red
and when you have `Test` active the header is for example blue? 

In that case you could use `useLocation` in the header and based on `pathname` decide what to do
```javascript
import { useLocation } from "@solidjs/router"

const location = useLocation()
const pathname = location.pathname
```

**Reply by ptasheq** [reply]
@hawkerboy7

Thanks for your answer. Yes, that's what I meant - you're right that I should have added something more tangible. I was thinking that maybe it can be done in a more data-driven way e.g.:


```javascript
export default function Pages() {
	return (<>
		<Route path="/" component={Home} rootProps={() => ({color: 'red'})} />
		<Route path="/test" component={Test} rootProps={() => ({color: 'blue'})} />
	</>)
}
```

Or maybe at least it's possible to access some context in `Header` to find out which `path` is active (as it's defined `/user/:id`, not as it's in the URL):
```javascript
const Header = () => {
	const {path} = useSomeContextExposedBySolidRouter();	
	const color = getColor(path);

	return <h2 class={`c-${color}`}>My App</h2>;
}
```

The solution using `useLocation` (which I appreciate) may not scale well because there'll be more than one source of truth and it'll become more complex for parametrized paths.

**Reply by hawkerboy7** [reply]
Alright I think it is clear.
I just started with SolidJS so don't know all the ins and outs, but perhaps [useContext](https://www.solidjs.com/docs/latest/api#usecontext) could be used for that purpose. You pass your `rootProps` to a self-created `generalProps-context` . Which you then access at any point in the application tree "below" the context definition
